pkg/tfbridge/info_defaults.go

```diff
@@ -0,0 +1,180 @@
+		}
+	}
+	return errs.ErrorOrNil()
+}
+
+type ProviderInfoDefaults struct {
+	Resources   func(resourceToken string) (*ResourceInfo, error)
+	DataSources func(dataSourceToken string) (*DataSourceInfo, error)
+}
+
+func (d ProviderInfoDefaults) Unmappable(substring string) ProviderInfoDefaults {
+	return ProviderInfoDefaults{
+		Resources:   ensureUnmappable(substring, d.Resources),
+		DataSources: ensureUnmappable(substring, d.DataSources),
+	}
+}
+
+func ensureUnmappable[T any](
+	substring string,
+	f func(t string) (*T, error),
+) func(t string) (*T, error) {
+	return func(t string) (*T, error) {
+		if strings.Contains(t, substring) {
+			return nil, fmt.Errorf("token %q matches a substring %q that was specified as Unmappable",
+				t, substring)
+		}
+		return f(t)
+	}
+}
+
+type TFToken struct {
+	Module string
+	Name   string
+}
+
+type TFTokenParser func(tfToken string) (*TFToken, error)
+
+func SingleModule(prefix, defaultModule string) TFTokenParser {
+	return KnownModules(prefix, defaultModule, nil)
+}
+
+func KnownModules(prefix, defaultModule string, modules []string) TFTokenParser {
+	// NOTE: We could turn this from a sort + linear lookup into a radix tree to recover
+	// O(log(n)) performance (current is O(n*m)) where n = number of modules and m =
+	// number of mappings.
+	sort.Sort(sort.Reverse(sort.StringSlice(modules)))
+
+	return func(tfToken string) (*TFToken, error) {
+		tk := strings.TrimPrefix(tfToken, prefix)
+		if len(tk) == len(tfToken) {
+			return nil, fmt.Errorf("token '%s' missing package prefix '%s'", tfToken, prefix)
+		}
+		mod := defaultModule
+		for _, m := range modules {
+			if strings.HasPrefix(tk, m) {
+				mod = m
+				break
+			}
+		}
+		if mod == "" {
+			return nil, fmt.Errorf("could not find a module that prefixes '%s' in '%#v'", tk, modules)
+		}
+		return &TFToken{
+			Module: mod,
+			Name:   strings.TrimPrefix(tk, mod),
+		}, nil
+	}
+}
+
+func SuggestTokens(pkg string, tfTokenParser TFTokenParser) ProviderInfoDefaults {
+	return ProviderInfoDefaults{
+		Resources: func(resourceToken string) (*ResourceInfo, error) {
+			tfTok, err := tfTokenParser(resourceToken)
+			if err != nil {
+				return nil, err
+			}
+			token := SuggestResourceToken(pkg, tfTok)
+			return &ResourceInfo{Tok: token}, nil
+		},
+		DataSources: func(dataSourceToken string) (*DataSourceInfo, error) {
+			tfTok, err := tfTokenParser(dataSourceToken)
+			if err != nil {
+				return nil, err
+			}
+			token := SuggestDataSourceToken(pkg, tfTok)
+			return &DataSourceInfo{Tok: token}, nil
+		},
+	}
+}
+
+func SuggestResourceToken(pkg string, token *TFToken) tokens.Type {
+	mod := camelCase(token.Module)
+	res := upperCamelCase(token.Name)
+	return MakeResource(pkg, mod, res)
+}
+
+func SuggestDataSourceToken(pkg string, token *TFToken) tokens.ModuleMember {
+	mod := camelCase(token.Module)
+	fn := upperCamelCase(token.Name)
+	submod := string(unicode.ToLower(rune(fn[0]))) + fn[1:]
+	return MakeMember(pkg, mod+"/"+submod, "get"+fn)
```

Comment:
So this bit I found on base branch I was surprised by. Is there a story for the "get" prefix? Is there a reason that MakeDataSource doesn't do the "get" prefix? Is this how it's conventionally done?